src/core/server/elasticsearch/version_check/ensure_es_version.ts

```diff
@@ -0,0 +1,151 @@
+  const incompatibleNodes = nodes.filter(
+    node => !isEsCompatibleWithKibana(node.version, kibanaVersion)
+  );
+
+  // Aggregate ES nodes which should prompt a Kibana upgrade. It's acceptable
+  // if ES and Kibana versions are not the same so long as they are not
+  // incompatible, but we should warn about it.
+  // Ignore version qualifiers https://github.com/elastic/elasticsearch/issues/36859
+  const warningNodes = nodes.filter(node => {
+    const nodeSemVer = coerce(node.version);
+    const kibanaSemver = coerce(kibanaVersion);
+    return nodeSemVer && kibanaSemver && nodeSemVer.version !== kibanaSemver.version;
+  });
+
+  let message;
+  if (incompatibleNodes.length > 0) {
+    const incompatibleNodeNames = incompatibleNodes.map(node => node.name).join(', ');
+    if (ignoreVersionMismatch) {
+      message = `Ignoring version incompatibility between Kibana v${kibanaVersion} and the following Elasticsearch nodes: ${incompatibleNodeNames}`;
+    } else {
+      message = `This version of Kibana (v${kibanaVersion}) is incompatible with the following Elasticsearch nodes in your cluster: ${incompatibleNodeNames}`;
+    }
+  } else if (warningNodes.length > 0) {
+    const warningNodeNames = warningNodes.map(node => node.name).join(', ');
+    message =
+      `You're running Kibana ${kibanaVersion} with some different versions of ` +
+      'Elasticsearch. Update Kibana or Elasticsearch to the same ' +
+      `version to prevent compatibility issues: ${warningNodeNames}`;
+  }
+
+  return {
+    isCompatible: ignoreVersionMismatch || incompatibleNodes.length === 0,
+    message,
+    incompatibleNodes,
+    warningNodes,
+    kibanaVersion,
+  };
+}
+
+export const pollEsNodesVersion = ({
+  callWithInternalUser,
+  log,
+  kibanaVersion,
+  ignoreVersionMismatch,
+  esVersionCheckInterval: healthCheckInterval,
+}: PollEsNodesVersionOptions) => {
+  log.debug('Checking Elasticsearch version');
+
+  return interval(healthCheckInterval).pipe(
+    switchMap(() => {
+      return callWithInternalUser('nodes.info', {
+        filterPath: ['nodes.*.version', 'nodes.*.http.publish_address', 'nodes.*.ip'],
+      });
+    }),
+    // Log, but otherwise ignore 'nodes.info' request errors
+    catchError((err, caught$) => {
+      log.error('Unable to retrieve version information from Elasticsearch nodes.', err);
+      return caught$;
+    }),
+    map((nodesInfo: NodesInfo) =>
+      mapNodesVersionCompatibility(nodesInfo, kibanaVersion, ignoreVersionMismatch)
+    ),
+    // Only emit if the IP or version numbers of the nodes changed from the
+    // previous result.
+    distinctUntilChanged((prev, curr) => {
+      const nodesEqual = (n: NodeInfo, m: NodeInfo) => n.ip === m.ip && n.version === m.version;
+      return (
+        curr.incompatibleNodes.every((node, i) => nodesEqual(node, prev.incompatibleNodes[i])) &&
+        curr.warningNodes.every((node, i) => nodesEqual(node, prev.warningNodes[i]))
+      );
```

Comment:
We also need to check the size of `incompatibleNodes` and `warningNodes`: if `curr.incompatibleNodes` is empty but `prev.incompatibleNodes` is not, it will not be flagged as changed.

Reply:
good catch